Backend/user improvements

cerob opened the pull request:
This PR includes:

- Some bug fixes after inspection with the team as mentioned in #118 
- Disallowing multiple registration with same email. (#123)
- Better handling of `CorporateUserProfile`

cerob commented:
Documentation #125 will be error free after this merge, so it's better to be quick.

mertakozcan commented:
I think UserCreateSerializer's create method and UserDetailsSerializer's update method do not handle tags properly. You can also fix that problem, it's handled in event serializers, you can copy from them. :)

mertakozcan commented on backend/api/serializers.py:
```diff
@@ -240,7 +240,7 @@ class Meta:
                   'following_count', 'owned_events_count', 'blocked_users_count',
                   'is_corporate_user', 'corporate_profile')
-        read_only_fields = ('username', 'tags', 'comments', 'votes',
+        read_only_fields = ('id', 'username', 'tags', 'comments', 'votes',
```
There is no need to specify `id` as read-only since it's already created as read-only by Django.

mertakozcan commented on backend/api/serializers.py:
```diff
@@ -249,9 +249,21 @@ def update(self, instance, validated_data):
+        else:
+            if instance.corporate_profile:
+                # case: corporate user profile is updated
+                instance.corporate_profile.url = corporate_profile.get('url', None)
```
```suggestion
                instance.corporate_profile.url = corporate_profile.get('url', None)
                instance.corporate_profile.save()
```

`instance.corporate_profile.save()` can be moved here and removed from line#265.

mertakozcan commented on backend/api/models.py:
```diff
@@ -107,6 +107,9 @@ class User(AbstractUser, CommentMixin, FollowMixin, TagMixin, VoteMixin):
                                              null=True, default=None)
 
+    class Meta:
+        unique_together = (('email',), ('username',))
```
🥇

cerob commented:
I implemented more:

- The changes @mertakozcan requested here.
- Added `profile_photo` field.
- Now we are handling user endpoint methods much better.
- Improved validations:
  - Disallow change of username.
  - Allow email change with unique constraint.
- Password change feature: 
  - `current_password` and `new_password` serializer fields.
  - Check for `current_password` is true.
  - Minimum 8, maximum 20 characters limitations validations as stated in the Requirements.
- Add tags update

mertakozcan commented on backend/api/serializers.py:
```diff
@@ -224,10 +230,57 @@ def create(self, validated_data):
+                serializers.ValidationError('Current password is wrong.')
+
+        # yet email can be changed
+        instance.email = validated_data.get('email', instance.email)
```
I think instead of explicitly setting all these fields, you can do something like

```
for key, value in validated_data.items():
    setattr(instance, key, value)
```

after you pop all the things from validated_data that requires manual handling. For example now, `profile_photo` isn't updated.

cerob commented:
As I know, we are not handling `tags` during sign up process. After the sign up verification, users are able to edit their profiles.

Other than that, I did small changes.

mertakozcan commented:
Okay then, thanks for your work!
